Add allclose example to docs

victorprincipe opened the pull request:
Here I have added examples to the `allclose.rst` file in the docs.

The examples are included from a python script (`allclose_example.py`) that has been added to the same directory as the `.rst` file.

Various examples have been included for the following methods:
- `allclose`
-  `allclose_raise`
- `allclose_block`
- `allclose_block_raise`

<!-- readthedocs-preview equistore start -->
----
:books: Documentation preview :books:: https://equistore--199.org.readthedocs.build/en/199/

<!-- readthedocs-preview equistore end -->

agoscinski commented on docs/src/reference/python/operations/logic/allclose_example.py:
```diff
@@ -0,0 +1,127 @@
+allclose_block(block1, block2)
+#Output:
+# False
+### This fails because allclose_block checks that the properties of two blocks are equal too
```
here I dont have a strong opinion, but maybe calling them `property_1` and `property_2` is easier to understand rather than one is upper case of the other and that is why they are different.

victorprincipe replied:
Bon idée

agoscinski commented on docs/src/reference/python/operations/logic/allclose_example.py:
```diff
@@ -0,0 +1,127 @@
+# Recreate block1, but change first value in the block from 1 to 1.00001
+block3 = TensorBlock(
+    values=np.array([
+         [1.00001, 2, 4],
```
Maybe it visually easier to grasp your examples below when you define 1.00001 as 1+1e-5

victorprincipe replied:
Très bon idée

agoscinski commented on docs/src/reference/python/operations/logic/allclose_example.py:
```diff
@@ -0,0 +1,127 @@
+# Using allclose_raise will raise a ValueError if the check fails, along with an explanation as to why 
+allclose_raise(tensor1, tensor4, rtol=1e-5)
+#Output:
+# ValueError: Inputs to allclose should have the same key indices.
```
Maybe for the `allclose_raises` one could write in the documentation here
https://github.com/lab-cosmo/equistore/blob/901fe071cc3c8ec87530d5669e8cf93fd8f2e611/python/src/equistore/operations/allclose.py#L51-L71
that it raises a ValueError if the keys are different, similar as in
https://github.com/numpy/numpy/blob/8cec82012694571156e8d7696307c848a7603b4e/numpy/linalg/linalg.py#L2201-L2204
for the convergence with
```
    Raises
    ------
    LinAlgError
        If computation does not converge.
```

victorprincipe replied:
In the description of the function, it already says to look at the `allclose` method for how two blocks are considered (not) equal. Please see below.

https://github.com/lab-cosmo/equistore/blob/901fe071cc3c8ec87530d5669e8cf93fd8f2e611/python/src/equistore/operations/allclose.py#L62-L64

agoscinski replied:
you are right, we don't need to repeat the information.

agoscinski reviewed (changes requested):
I think it is already looking good, but one important change that needs to be done is to move the examples to the corresponding functions `allclose`, `allclose_block` and `allclose_block_raise`, `allclose_raise` as it is done for the `reduce_over_samples` function here
https://github.com/lab-cosmo/equistore/blob/16ff8ea4b1b466892473e53bf042e1683417e410/python/src/equistore/operations/reduce_over_samples.py#L266-L327
In the doc it looks then like this
https://lab-cosmo.github.io/equistore/latest/reference/python/operations/manipulation/samples-reduction.html#equistore.sum_over_samples

This has the advantage that it is in the help page of the function using `help` and we can automatically test if the doc test is printing the same output as specified in the doc with `pytest --doctest-modules --pyargs equistore`

DavideTisi commented:
Very cool, one thing in the doc the actual docstrings are not shown anymore

victorprincipe commented:
I've added the examples to the functions themselves, and removed the examples section from the `.rst` file for the doc. Please let me know what you think.

Also, in response to this @DavideTisi:

> Very cool, one thing in the doc the actual docstrings are not shown anymore

Seems like they are quoted here: https://lab-cosmo.github.io/equistore/latest/reference/python/operations/manipulation/samples-reduction.html

victorprincipe commented:
Also, maybe we should run some other tests apart from just building the docs, as I altered module files.

victorprincipe commented:
> Very cool, one thing in the doc the actual docstrings are not shown anymore

Ah sorry, understand what you meant now. This was an issue with tax I believe which was fixed in the master branch. I have rebased and now it should work.

agoscinski commented on python/src/equistore/operations/allclose.py:
```diff
@@ -63,6 +149,71 @@ def allclose_raise(
+    >>> tensor2 = TensorMap(keys, [block2])
+    ...
+    >>> # Call allclose_raise, which should return a ValueError because:
+    >>> # 1. The two NaNs are not considered equal of the two tensors is 1e-5,
```
Should it be like this? I don't understand the current sentence
```suggestion
    >>> # 1. The two NaNs are not considered equal,
```

victorprincipe replied:
Yes, typo, will commit the suggestion

agoscinski commented on python/src/equistore/operations/allclose.py:
```diff
@@ -63,6 +149,71 @@ def allclose_raise(
+    >>> # Call allclose_raise, which should return a ValueError because:
+    >>> # 1. The two NaNs are not considered equal of the two tensors is 1e-5,
+    >>> # 2. The difference between the first value in the blocks
+    >>> #    which is greater than the default rtol of 1e-13
```
Should it be like this? "The difference [...] is greater than [...]"
```suggestion
    >>> # is greater than the default rtol of 1e-13
```

victorprincipe replied:
Yes, will also commit this.

agoscinski reviewed (changes requested):
Looks almost ready to squash and rebase to me, just two sentences I did not quite get.

agoscinski reviewed (approved):
I think it looks good now
